tcp.c

```diff
@@ -418,91 +419,71 @@ tcp_tls_get_server_pubkey(STREAM s)
 
 /* Establish a connection on the TCP layer */
 RD_BOOL
-tcp_connect(char *server)
+tcp_connect(struct addrinfo *ai)
 {
-	socklen_t option_len;
+	int i, res;
 	uint32 option_value;
-	int i;
-
-#ifdef IPv6
-
-	int n;
-	struct addrinfo hints, *res, *ressave;
-	char tcp_port_rdp_s[10];
-
-	snprintf(tcp_port_rdp_s, 10, "%d", g_tcp_port_rdp);
+	char hostname[64];
+	struct addrinfo *pai;
+	socklen_t option_len;
 
-	memset(&hints, 0, sizeof(struct addrinfo));
-	hints.ai_family = AF_UNSPEC;
-	hints.ai_socktype = SOCK_STREAM;
+	/* connect to host */
+	pai = ai;
 
-	if ((n = getaddrinfo(server, tcp_port_rdp_s, &hints, &res)))
-	{
-		logger(Core, Error, "tcp_connect(), getaddrinfo() failed: %s", gai_strerror(n));
-		return False;
-	}
+	/* if we have a previous used address, then only try
+	   connecting to this specific address. */
+	if (connection_pai != NULL)
+		pai = connection_pai;
 
-	ressave = res;
 	g_sock = -1;
-	while (res)
+	while (pai)
 	{
-		g_sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
-		if (!(g_sock < 0))
+		g_sock = socket(pai->ai_family, pai->ai_socktype, pai->ai_protocol);
+		if (g_sock > 0)
 		{
-			if (connect(g_sock, res->ai_addr, res->ai_addrlen) == 0)
-				break;
-			TCP_CLOSE(g_sock);
-			g_sock = -1;
-		}
-		res = res->ai_next;
-	}
-	freeaddrinfo(ressave);
+			/* try connect to server */
+			getnameinfo(pai->ai_addr,  pai->ai_addrlen, hostname, sizeof(hostname), NULL, 0, NI_NUMERICHOST);
 
-	if (g_sock == -1)
-	{
-		logger(Core, Error, "tcp_connect(), unable to connect to %s", server);
-		return False;
-	}
+			if (connection_pai == pai)
+				logger(Core, Debug, "tcp_connect(), reconnecting to %s", hostname);
+			else
+				logger(Core, Debug, "tcp_connect(), connecting to %s", hostname);
 
-#else /* no IPv6 support */
+			res = connect(g_sock, pai->ai_addr, pai->ai_addrlen);
+			if (res == 0)
+			{
+				/* We are connected, lets store addrinfo pointer of
+				   connection and break out of connect loop */
+				connection_pai = pai;
+				break;
+			}
 
-	struct hostent *nslookup;
-	struct sockaddr_in servaddr;
+			/* Failed to reconnect to previous connected host, lets fail */
+			if (connection_pai == pai)
+			{
+				pai = NULL;
+				break;
+			}
```

Comment:
Don't we want to keep this if we're trying to reconnect after a network error? We might get a few failed `connect()` before we get back in touch with the server.

Reply:
`pai` is local variable and only used to instrument a connect failure further down, whats important here is the connection_pai which is the pointer to the connection to be made which only updates on successful connection.